### PR TITLE
Use onBecomeUnobserved for ObservableMap._hasMap garbage collection

### LINKED ISSUE #2031: Memory leaking of Observable Maps (), because _hasMap keeps links on all once used keys (key could be object) and never delete them
Issue:
I find that observable Maps absolutely always loose a memory. For every key of the map, you create same key in your private map called _hasMap and place there a boolean value. But these keys are kept there even the origin map is empty or already doesn't contain the key. Don't forget that key could be large object or linked with many other objects. So we loose all memory even we don't keep links (have variables) to these key object in our code, because the _hasMap keeps links on all once used keys (key - is the big object, remember ?) and never delete them.

Please see the example:
https://codesandbox.io/s/minimal-mobx-react-project-1uyb0

Of cause we expected that we don't see the our objects in memory heap, after clearing the map.

Idea:
You should use another mechanism to register of using the map methods and of getting values.

### PR DESCRIPTION
**Pull request by mayorovp**
fixes #2031

* [x] Added unit tests
* [ ] Updated changelog
* ~Updated docs (either in the description of this PR as markdown, or as separate PR on the `gh-pages` branch. Please refer to this PR). For new functionality, at least [API.md](https://github.com/mobxjs/mobx/blob/gh-pages/docs/refguide/api.md) should be updated~
* ~Added typescript typings~
* [ ] Verified that there is no significant performance drop (`npm run perf`)

### CODE REVIEW

**Comment by FredyC** [comment]
@mayorovp Please fix tests if you can. Also what about adding some additional tests to ensure we won't break in the future?

And we will need to port this to `mobx4-master` eventually, but looks like it shouldn't be such a problem.

**Comment by mayorovp** [comment]
@FredyC I don't know why `getAtom()` return `has` atom when no `data` atom is present, so I don't know how fix this error.

**Comment by mweststrate** [comment]
I'll try to dive into it early next week. It might also be solvable without the atoms, if `onBecomeOpbserved(this.hasMap, key, callback)` is used

File: test/base/map.js [hunk 1]
```diff
@@ -227,7 +227,28 @@ test("cleanup", function() {
     disposer()
     expect(aValue).toBe(2)
     expect(observable.observers.size).toBe(0)
-    expect(x._hasMap.get("a").observers.size).toBe(0)
+    expect(x._hasMap.get("a")).toBe(undefined)
+})
+
+test("getAtom encapsulation leak test", function() {
```
Comment (BanderasPRO) [inline test/base/map.js]: Please add a test, to check existence of a key in the _hasMap, not only empty value of the key

Comment (mayorovp) [reply test/base/map.js]: `Map<K, ObservableValue<boolean>>` cannot contain empty values.

Comment (BanderasPRO) [reply test/base/map.js]: I'm not sure that this condition of definition works in runtime. 
Native JS environment allows that.
you can just check the _hasMap size

Comment (mayorovp) [reply test/base/map.js]: Of course it works except when some typehacks broke all type safety.

But I already changed this line.

**Comment by mweststrate** [comment]
Looking good to me as well, merging!

**Comment by mweststrate** [comment]
Released as 4.13.0 / 5.13.0
